networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/peers/Eth2PeerSelectionStrategy.java

```diff
@@ -38,54 +47,149 @@
   private final TargetPeerRange targetPeerCountRange;
   private final PeerSubnetSubscriptions.Factory peerSubnetSubscriptionsFactory;
   private final ReputationManager reputationManager;
+  private final Shuffler shuffler;
 
   public Eth2PeerSelectionStrategy(
       final TargetPeerRange targetPeerCountRange,
       final PeerSubnetSubscriptions.Factory peerSubnetSubscriptionsFactory,
-      final ReputationManager reputationManager) {
+      final ReputationManager reputationManager,
+      final Shuffler shuffler) {
     this.targetPeerCountRange = targetPeerCountRange;
     this.peerSubnetSubscriptionsFactory = peerSubnetSubscriptionsFactory;
     this.reputationManager = reputationManager;
+    this.shuffler = shuffler;
   }
 
   @Override
   public List<PeerAddress> selectPeersToConnect(
-      final P2PNetwork<?> network, final Supplier<List<DiscoveryPeer>> candidates) {
+      final P2PNetwork<?> network,
+      final PeerSources peerSources,
+      final Supplier<List<DiscoveryPeer>> candidates) {
     final PeerSubnetSubscriptions peerSubnetSubscriptions =
         peerSubnetSubscriptionsFactory.create(network);
     final int peersRequiredForPeerCount =
         targetPeerCountRange.getPeersToAdd(network.getPeerCount());
+    final int randomlySelectedPeerCount = getCurrentRandomlySelectedPeerCount(network, peerSources);
+    final int randomlySelectedPeersToAdd =
+        targetPeerCountRange.getRandomlySelectedPeersToAdd(randomlySelectedPeerCount);
+
     final int peersRequiredForSubnets = peerSubnetSubscriptions.getSubscribersRequired();
-    final int maxAttempts = Math.max(peersRequiredForPeerCount, peersRequiredForSubnets);
-    LOG.trace("Connecting to up to {} known peers", maxAttempts);
-    if (maxAttempts == 0) {
+    final int scoreBasedPeersToAdd =
+        Math.max(peersRequiredForPeerCount - randomlySelectedPeerCount, peersRequiredForSubnets);
+    LOG.trace(
+        "Connecting to up to {} known peers", scoreBasedPeersToAdd + randomlySelectedPeersToAdd);
+    if (scoreBasedPeersToAdd == 0 && randomlySelectedPeersToAdd == 0) {
       return emptyList();
     }
+
+    final List<DiscoveryPeer> allCandidatePeers = new ArrayList<>(candidates.get());
+    final List<PeerAddress> selectedPeers = new ArrayList<>();
+
+    if (randomlySelectedPeersToAdd > 0) {
+      selectedPeers.addAll(
+          selectAndRemoveRandomPeers(
+              network, peerSources, randomlySelectedPeersToAdd, allCandidatePeers));
```

Comment:
Yeah I was in two minds because it removes the invalid peers it finds as well as the ones it selects, but it's probably better to keep it simple and just wind up re-evaluating the rejected peers again when we consider them based on score.

Reply:
Ah, just remembered why I wound up doing this - there's a conversion from `DiscoveryPeer` to `PeerAddress` required to check if the candidate peer is already connected so having the chosen `PeerAddress` instances isn't enough to let us remove the `DiscoveryPeer` instances.  I don't like it but haven't come up with a way around it yet.